File: Leetcode/Top Interview Questions/17.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        characters = {'2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
                      '6': 'mon', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'}

        if len(digits) == 0:
            return []
        if len(digits) == 1:
            return list(characters[digits[0]])
        prev = self.letterCombinations(digits[:-1])
        additional = characters[digits[-1]]
        return [s + c for s in prev for c in additional]

File: Leetcode/Recursion/17.js
/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function (digits) {
  if (digits.length === 0) return [];

  let answer = [];
  const letters = {
    2: "abc",
    3: "def",
    4: "ghi",
    5: "jkl",
    6: "mno",
    7: "pqrs",
    8: "tuv",
    9: "wxyz",
  };

  function makeCombination(count, s) {
    if (count === digits.length) return answer.push(s);

    for (letter of letters[digits[count]]) {
      makeCombination(count + 1, s + letter);
    }
  }

  makeCombination(0, "");

  return answer;
};
